tests/test_cli.py

```diff
@@ -1,152 +1,218 @@
 """Tests for the CLI."""
 
 from pathlib import Path
+from unittest import mock
 
 import pytest
+from click.testing import CliRunner
 
-from nipoppy.cli.run import cli
+from nipoppy.cli import cli
 
 from .conftest import ATTR_TO_DPATH_MAP
 
+runner = CliRunner()
+
 
 def test_cli():
-    with pytest.raises(SystemExit) as exception:
-        cli(["nipoppy", "-h"])
-    assert exception.value.code == 0
+    result = runner.invoke(
+        cli,
+        ["-h"],
+    )
+    assert result.exit_code == 0
 
 
 def test_cli_invalid():
-    with pytest.raises(SystemExit) as exception:
-        cli(["nipoppy", "--fake-arg"])
-    assert exception.value.code != 0
+    result = runner.invoke(
+        cli,
+        ["--fake-arg"],
+    )
+    assert result.exit_code != 0
 
 
 def test_cli_init(tmp_path: Path):
-    try:
-        cli(["nipoppy", "init", str(tmp_path / "my_dataset")]) is None
-    except SystemExit:
-        pass
+    result = runner.invoke(
+        cli,
+        ["init", "--dataset", str(tmp_path / "my_dataset")],
+    )
+    assert result.exit_code == 0
+
+
+def test_cli_init_dir_exists(tmp_path: Path):
+    dpath_root = tmp_path / "my_dataset"
+    dpath_root.joinpath("dir_exists").mkdir(parents=True, exist_ok=True)
+
+    with pytest.raises(FileExistsError):
+        runner.invoke(
+            cli,
+            ["init", "--dataset", str(dpath_root)],
+            catch_exceptions=False,
+        )
 
 
 def test_cli_doughnut(tmp_path: Path):
     dpath_root = tmp_path / "my_dataset"
-    try:
-        cli(["nipoppy", "doughnut", str(dpath_root)])
-    except BaseException:
-        pass
-
-    # check that a logfile was created
-    assert (
-        len(list((dpath_root / ATTR_TO_DPATH_MAP["dpath_logs"]).glob("doughnut/*.log")))
-        == 1
-    )
+    with pytest.raises(RuntimeError) as excinfo, mock.patch("sys.exit") as mock_exit:
+        result = runner.invoke(
+            cli,
+            ["doughnut", "--dataset", str(dpath_root)],
+            catch_exceptions=False,
+        )
+        mock_exit.assert_called_once_with(1)
+
+        # check that a logfile was created
+        assert (
+            len(
+                list(
+                    (dpath_root / ATTR_TO_DPATH_MAP["dpath_logs"]).glob(
+                        "doughnut/*.log"
+                    )
+                )
+            )
+            == 1
+        )
+
+        # Dataset was not initialized properly
+        assert excinfo.value.startswtih(
+            "Dataset does not follow expected directory structure:"
+        )
+        assert result.exit_code == 1
```

Comment:
I don't think these lines are being run at all because the exception is thrown (e.g. there is a typo in `startswtih` but the test still passes). So we would need to de-indent some of this. But that still doesn't fix the coverage things though I have an idea for that, which would also address another comment I had about exceptions being written to the logger.

In `cli.py` we could have something like that for the subcommands:
```python
from nipoppy.env import ReturnCode

def doughnut(**params):
    """Command: nipoppy doughnut."""
    from nipoppy.workflows.doughnut import DoughnutWorkflow

    workflow = DoughnutWorkflow(**params)
    try:
        workflow.run()
    except Exception:
        workflow.logger.exception("Error while running nipoppy")
        if workflow.return_code == ReturnCode.SUCCESS:
            workflow.return_code = 2  # we could make a ReturnCode.OTHER_FAILURE for that

    sys.exit(workflow.return_code)
```

Then in `test_cli.py` the test could be like:
```python
def test_cli_doughnut(tmp_path: Path):
    dpath_root = tmp_path / "my_dataset"

    result = runner.invoke(cli, ["doughnut", "--dataset", str(dpath_root)])

    # check that a logfile was created
    assert (
        len(list((dpath_root / ATTR_TO_DPATH_MAP["dpath_logs"]).glob("doughnut/*.log")))
        == 1
    )
    assert result.exit_code == 2  # or ReturnCode.OTHER_FAILURE if we decide to do that
```

There is still a remaining thing that I think is suboptimal compared to what I had before: the logger won't catch errors that occur during workflow instantiation. For example, if I run `nipoppy init --dataset . -vvvv` (too many verbosity counts) then I get an ugly traceback with all the `click`-related frames. Could be sort-of "fixed" by not doing anything other than variable assignment in `__init__()`. But that is probably why I was passing an already-instantiated logger to the workflow classes instead of letting them make their own